awx/main/tasks/host_metrics.py

```diff
@@ -3,33 +3,80 @@
+
+
+@task(queue=get_task_queuename)
+def cleanup_host_metrics():
+    if is_run_threshold_reached(getattr(settings, 'CLEANUP_HOST_METRICS_LAST_TS', None), getattr(settings, 'CLEANUP_HOST_METRICS_INTERVAL', 30) * 86400):
+        logger.info(f"Executing cleanup_host_metrics, last ran at {getattr(settings, 'CLEANUP_HOST_METRICS_LAST_TS', '---')}")
+        HostMetricTask().cleanup(
+            soft_threshold=getattr(settings, 'CLEANUP_HOST_METRICS_SOFT_THRESHOLD', 12),
+            hard_threshold=getattr(settings, 'CLEANUP_HOST_METRICS_HARD_THRESHOLD', 36),
+        )
+        logger.info("Finished cleanup_host_metrics")
 
 
 @task(queue=get_task_queuename)
 def host_metric_summary_monthly():
     """Run cleanup host metrics summary monthly task each week"""
-    if _is_run_threshold_reached(
-        getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', None), getattr(settings, 'HOST_METRIC_SUMMARY_TASK_INTERVAL', 7) * 86400
-    ):
+    if is_run_threshold_reached(getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', None), getattr(settings, 'HOST_METRIC_SUMMARY_TASK_INTERVAL', 7) * 86400):
         logger.info(f"Executing host_metric_summary_monthly, last ran at {getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', '---')}")
         HostMetricSummaryMonthlyTask().execute()
         logger.info("Finished host_metric_summary_monthly")
 
 
-def _is_run_threshold_reached(setting, threshold_seconds):
-    last_time = DateTimeField().to_internal_value(setting) if setting else DateTimeField().to_internal_value('1970-01-01')
+class HostMetricTask:
+    """
+    This class provides cleanup task for HostMetric model.
+    There are two modes:
+    - soft cleanup (updates columns delete, deleted_counter and last_deleted)
+    - hard cleanup (deletes from the db)
+    """
+
+    def cleanup(self, soft_threshold=None, hard_threshold=None):
+        """
+        Main entrypoint, runs either soft cleanup, hard cleanup or both
 
-    return (now() - last_time).total_seconds() > threshold_seconds
+        :param soft_threshold: (int)
+        :param hard_threshold: (int)
+        """
+        if soft_threshold is not None:
+            self.soft_cleanup(soft_threshold)
+        if hard_threshold is not None:
+            self.hard_cleanup(hard_threshold)
+
+        settings.CLEANUP_HOST_METRICS_LAST_TS = now()
+
+    @staticmethod
+    def soft_cleanup(threshold=None):
+        if not threshold:
+            threshold = getattr(settings, 'CLEANUP_HOST_METRICS_SOFT_THRESHOLD', 12)
+
+        last_automation_before = now() - relativedelta(months=threshold)
+        logger.info(f'cleanup_host_metrics: soft-deleting records last automated before {last_automation_before}')
+        HostMetric.active_objects.filter(last_automation__lt=last_automation_before).update(
+            deleted=True, deleted_counter=F('deleted_counter') + 1, last_deleted=now()
+        )
+
+    @staticmethod
+    def hard_cleanup(threshold=None):
+        if not threshold:
```

Comment:
The logic of `if not threshold` is blocking me from running

```
awx-manage cleanup_host_metrics --soft 0
```

to test this PR, because `0` is a falsy value.

Reply:
it's not specified in the ADR that soft delete could have the meaning "delete all". 
My understanding is, that the sense of this feature is different, so this command from my point of view is invalid

Reply:
@Ladas ^ what's your thought?

Reply:
I don't fully know what your reasons are for not allowing "delete all", but using a value of `0` months is going to be first and most obvious approach to testing. If conflicts with objectives you have, then I don't need to give pushback, we can hack around it for testing.

But if this command doesn't allow `0` months retention, silently doing nothing when given the `0` value is not the right behavior. You could throw a command error in that case. Exact implementation is up to you, I just don't want it to work in a frustrating and unexpected way.

Reply:
I got an answer that it's not wanted to allow configurable command. It'll be driven by settings only.

Reply:
iiuc, you can probably get what Alan and you both want with, *explicitly*, `if threshold == None` <-- This explicitly says that the value was not passed in. You should be doing this anyway from a readability perspective because you're using `None` to indicate that the value was not passed in. If you just say `not threshold` then you'll catch a lot of situations where the value *WAS* passed in: e.g., `myfunc(threshold=0)`, `myfunc(threshold=False`, etc.

Reply:
now it receives 0 or "0", raises exceptions on "", set() etc.